### PR TITLE
VRR mode (G-Sync, FreeSync).

### PR DESCRIPTION
**Pull request by Tatsuya79**
## Description

Variable refresh rate screens aren't keeping a smooth rendering in RA atm.
Testing on a G-Sync compatible monitor with an Nvidia card, only forcing max speed to x1.0 (in Frame Throttle Menu) ~~and also disabling Audio Sync (Audio Menu)~~ would give me a constantly smooth behaviour.
But in this configuration you're losing the possibility to fast forward.

This PR tries to address those issue:
-forcing x1
-disabling Audio Sync
-adding a way to fast forward in that mode

edit:
Max speed x1 is enough to get a smooth result in RA.  
It's **after fast forwarding with my modifications** that it stutters and disabling audio sync allows the animation to stay smooth in that scenario. 
So I'm perhaps getting outside of the right frame timing by fast forwarding the way I did here...
Or some calculations that take the monitor refresh rate  into account are playing there.

## Related Issues

Audio will crackle if you can't meet the fps timing requirement.

Mednafen PSX has strange timing variations that will make the sound crackle in that mode.

Cores changing their refresh rate during execution while not calling RETRO_ENVIRONMENT_SET_SYSTEM_AV_INFO won't see the new timing applied. (Dolphin when booting some EUR games that ask for a 60hz switch later)
This kind of issue was already happening before when running with max speed x1. (I fixed that for px68k)

Any way to make the sound adapt to loss of framerate/timing issues without crackling?

## Reviewers

@Themaister will know if I'm calling stuff I shouldn't in the bad places.

### CODE REVIEW

File: retroarch.c [hunk 1]
```diff
@@ -1755,8 +1755,9 @@ bool rarch_ctl(enum rarch_ctl_state state, void *data)
                ? 1.0f : settings->floats.fastforward_ratio;
 
             frame_limit_last_time    = cpu_features_get_time_usec();
+
             frame_limit_minimum_time = (retro_time_t)roundf(1000000.0f
-                  / (av_info->timing.fps * fastforward_ratio));
+            / (av_info->timing.fps * (settings->bools.vrr_runloop_enable ? 1.0f : fastforward_ratio)));
```
[the remark this answers is not in the document]
Comment (Tatsuya79) [reply retroarch.c]: It's applied later in runloop_iterate after l.3408.

**Comment by Tatsuya79** [comment]
What makes me disable audio blocking is that after doing fast forward or rewind there will be a lot of stutters.
Perhaps it could be avoided in a better way?
I tried many things without success.

File: audio/audio_driver.c [hunk 2]
```diff
@@ -484,7 +484,8 @@ static bool audio_driver_init_internal(bool audio_cb_inited)
          && current_audio->use_float(audio_driver_context_audio_data))
       audio_driver_use_float = true;
 
-   if (!settings->bools.audio_sync && audio_driver_active)
+   /* Disable audio sync for vrr screens as it induces frame timing deviation. */
+   if ((!settings->bools.audio_sync || settings->bools.vrr_runloop_enable) && audio_driver_active)
```
[the remark this answers is not in the document]
Comment (Tatsuya79) [reply audio/audio_driver.c]: Perhaps it's not the right answer, yes. I updated the description to explain why I came to do that.

**Comment by bparker06** [comment]
>audio resampling is setup to inherit the game refresh rate and not configured display refresh rate

did you try this?

**Comment by Tatsuya79** [comment]
[I'm doing it like that now (not touching audio_driver.c).](https://pastebin.com/PATKptGJ)
It's working fine for me but I need confirmation from others (if that's still smooth after fast forwarding).

I'll close this one and send a new PR if that's confirmed and is acceptable (I'm bad with git).

**Comment by twinaphex** [comment]
All fine with me. Hopefully we can get some people testing this now like @bparker06 .

**Comment by Tatsuya79** [comment]
Closing that one.

**Pull request closed without merging**